modules/frontend/search_streaming.go

```diff
@@ -87,23 +87,104 @@ func (p *diffSearchProgress) result() *shardedSearchResults {
 	return res
 }
 
+func (p *diffSearchProgress) metrics() *tempopb.SearchMetrics {
+	return p.progress.metrics()
+}
+
 // finalResult gives the user the ability to pull all results w/o filtering
 // to ensure that all results are sent to the caller
 func (p *diffSearchProgress) finalResult() *shardedSearchResults {
 	return p.progress.result()
 }
 
+type multiProgress struct {
+	mu              sync.Mutex
+	progress        []*diffSearchProgress
+	combinedMetrics *tempopb.SearchMetrics
+}
+
+func newMultiProgress() *multiProgress {
+	return &multiProgress{
+		progress:        make([]*diffSearchProgress, 0),
+		combinedMetrics: &tempopb.SearchMetrics{},
+	}
+}
+
+func (mp *multiProgress) Add(p *diffSearchProgress) {
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+
+	mp.progress = append(mp.progress, p)
+
+	// combine metrics
+	m := *p.metrics()
+	// only set the metrics we set in progress constructor, other metrics are set by the sharder
+	mp.combinedMetrics.TotalBlocks += m.TotalBlocks
+	mp.combinedMetrics.TotalBlockBytes += m.TotalBlockBytes
+	mp.combinedMetrics.TotalJobs += m.TotalJobs
+}
+
+// results will calls result on underlying progress objects and combines them into a single result.
+func (mp *multiProgress) results(ctx context.Context) *shardedSearchResults {
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+
+	if len(mp.progress) == 1 {
+		// single tenant query, only one progress object.
+		return mp.progress[0].result()
+	}
+
+	comb := newSearchProgress(ctx, 0, mp.combinedMetrics.TotalJobs, mp.combinedMetrics.TotalBlocks, mp.combinedMetrics.TotalBlockBytes)
+	for _, progress := range mp.progress {
+		r := *progress.result() // only get new results from each progress object
```

Comment:
i don't think there's a reason to dereference

Reply:
yup, cleaned up.